agent/consul/leader_routine_manager.go

```diff
@@ -0,0 +1,120 @@
+package consul
+
+import (
+	"context"
+	"log"
+	"os"
+	"sync"
+)
+
+type LeaderRoutine func(ctx context.Context) error
+
+type leaderRoutine struct {
+	running bool
+	cancel  context.CancelFunc
+}
+
+type LeaderRoutineManager struct {
+	lock   sync.RWMutex
+	logger *log.Logger
+
+	routines map[string]*leaderRoutine
+}
+
+func NewLeaderRoutineManager(logger *log.Logger) *LeaderRoutineManager {
+	if logger == nil {
+		logger = log.New(os.Stderr, "", log.LstdFlags)
+	}
+
+	return &LeaderRoutineManager{
+		logger:   logger,
+		routines: make(map[string]*leaderRoutine),
+	}
+}
+
+func (m *LeaderRoutineManager) IsRunning(name string) bool {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if routine, ok := m.routines[name]; ok {
+		return routine.running
+	}
+
+	return false
+}
+
+func (m *LeaderRoutineManager) Start(name string, routine LeaderRoutine) error {
+	return m.StartWithContext(nil, name, routine)
+}
+
+func (m *LeaderRoutineManager) StartWithContext(parentCtx context.Context, name string, routine LeaderRoutine) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if instance, ok := m.routines[name]; ok && instance.running {
+		return nil
+	}
+
+	if parentCtx == nil {
+		parentCtx = context.Background()
+	}
+
+	ctx, cancel := context.WithCancel(parentCtx)
+	instance := &leaderRoutine{
+		running: true,
+		cancel:  cancel,
+	}
+
+	go func() {
+		err := routine(ctx)
```

Comment:
This is totally fine for the reason you mentioned (it has to gain the overall lock with the method that just spawned this is already holding) but additionally because the go routine specifically only ever touches the `*leaderRoutine` struct. One previous implementation didn't use a separate struct with a pointer to it and there were still some issues with a quick stop then start as we are not waiting for any of the routines to finish. It was overwriting the status and things got really messed up. Moving to the pointer to the struct ensured that every routine run has its own tracking data and that when we do a stop we just unlink that associated data after cancelling it. That way when a new instance gets started they cannot possibly overwrite each others status.